### DISCUSSION TITLE
keycloak 19 connect to postman

### DISCUSSION THREAD

**Discussion by jasondee1992** [keycloak/keycloak#15417] [Q&A]
Trying to connect to admin api using http://localhost:8080/admin/realms in postma but I received an error. Can I ask on how to connect my keycloak 19 to postman ?
![image](https://user-images.githubusercontent.com/49851974/200789249-fe89a6ae-3e1f-401f-b134-52e72f671b8f.png)

**Comment by souravs17031999** [comment]
@jasondee1992 , All these admin route API's are authorized routes in Keycloak and needs an Authorization header to be sent as Bearer token.

You will need to get the master user token as shown below : 

![image](https://user-images.githubusercontent.com/33771969/200850734-c7ee1f70-5565-4b7d-aa07-01a1f7cd211f.png)

Then, use this as bearer token in postman to send the request as shown below : 
![image](https://user-images.githubusercontent.com/33771969/200851328-0e112459-d7f5-4e10-b323-093b32fa17bc.png)


Also, I see you are using POST request to /admin/realms , I suspect that would't work without correct body. For this consult, API docs but anyways it should work with above.

**Reply by jasondee1992** [reply]
Trying the bearer token but I can't connect
![image](https://user-images.githubusercontent.com/49851974/200972150-926c40f9-121b-44e0-a900-88fd352b89a6.png)
![image](https://user-images.githubusercontent.com/49851974/200972291-898f900a-ac43-4a32-887d-587e533251fc.png)

![image](https://user-images.githubusercontent.com/49851974/200972231-eff08a9c-eb1e-4632-b3f2-6ef2f4471f2a.png)

**Reply by souravs17031999** [reply]
@jasondee1992 You need to use "admin-cli" client which I showed you earlier to generate the token using "client_credentials" grant and use that token for invoking the REST admin API's of Keycloak.
Client Initial acess token which you are showing will not work here for admin API's since they have a different purpose (they are used for OIDC dynamic client registration).

**Reply by jasondee1992** [reply]
I can't generate token in admin-cli ?



![image](https://user-images.githubusercontent.com/49851974/201003325-4214f8c3-ebf7-4c56-9e51-7daf3adb792e.png)

**Reply by souravs17031999** [reply]
Generate it via Postman (as shown earlier) , I don't think there's anyway to generate using the Admin console.

**Reply by jasondee1992** [reply]
Sorry sir I didn't understand. I'm new in keycloak Trying to follow your instruction in http://localhost:8080/auth/realms/master/protocol/openid-connect/token link but I received this. I can't receive in the body.  Do I need to setup first my keycloak before I can connect to using postman ? 

I run my keycloak using running CMD keycloak folder path bin>kc.bat start-dev

![image](https://user-images.githubusercontent.com/49851974/201041434-75ea4b44-feb0-4cc3-84dc-47e37d4f412b.png)
![image](https://user-images.githubusercontent.com/49851974/201044144-0d9dc202-1084-4ff0-8c27-fa737d730f4d.png)

**Reply by souravs17031999** [reply]
The URL in second approach (in pic) is correct, you just need to set the body in postman (choose x-www-url-encoded and pass the required keys/values) as I showed above, It will work.
If you will send the request without body, it will throw "unknown_error" 500 as It's currently.

**Reply by lexcao** [reply]
@jasondee1992 
From the screenshot you provide, I can see you're not putting the body data in your request.

If there is a body, there should be a green point, like this.
<img width="720" alt="image" src="https://user-images.githubusercontent.com/14901139/201140176-e383307c-d0bf-4e64-a5e4-1b727681e3c1.png">

Could you please try this cURL?
```
curl --request POST \
  --url http://127.0.0.1:8080/realms/master/protocol/openid-connect/token \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data client_id={} \
  --data client_secret={} \
  --data grant_type=password \
  --data username={} \
  --data password={}
```

**Reply by jasondee1992** [reply]
> @jasondee1992 From the screenshot you provide, I can see you're not putting the body data in your request.
> 
> If there is a body, there should be a green point, like this. <img alt="image" width="720" src="https://user-images.githubusercontent.com/14901139/201140176-e383307c-d0bf-4e64-a5e4-1b727681e3c1.png">
> 
> Could you please try this cURL?
> 
> ```
> curl --request POST \
>   --url http://127.0.0.1:8080/realms/master/protocol/openid-connect/token \
>   --header 'Content-Type: application/x-www-form-urlencoded' \
>   --data client_id={} \
>   --data client_secret={} \
>   --data grant_type=password \
>   --data username={} \
>   --data password={}
> ```

Thank you sir. I got this answer.
